### PR TITLE
(std/file_server): better template

### PR DESCRIPTION
**Pull request by zhmushan**
image
image

### CODE REVIEW

File: std/http/file_server.ts [hunk 1]
```diff
@@ -126,7 +81,7 @@ async function serveFile(
   const fileInfo = await stat(filePath);
   const headers = new Headers();
   headers.set("content-length", fileInfo.len.toString());
-  headers.set("content-type", contentType(extname(filePath)) || "text/plain");
+  headers.set("content-type", "text/plain");
```
Comment (kevinkassimo) [inline std/http/file_server.ts]: Is this the behavior we want?

Comment (zhmushan) [reply std/http/file_server.ts]: If using `contentType()`, sometimes the browser will download the file directly

Comment (kevinkassimo) [reply std/http/file_server.ts]: @zhmushan Yeah but that was my intention... I remember Python `SimpleHTTPServer` similarly set content types

Comment (zhmushan) [reply std/http/file_server.ts]: I think we prefer to open a typescript file instead of downloading it

Comment (zhmushan) [reply std/http/file_server.ts]: > @zhmushan Yeah but that was my intention... I remember Python `SimpleHTTPServer/http.server` similarly set content types
> Ref: https://github.com/python/cpython/blob/305189ecdc8322c22879a04564cad5989f937462/Lib/http/server.py#L694, https://github.com/python/cpython/blob/305189ecdc8322c22879a04564cad5989f937462/Lib/http/server.py#L854-L876

It is a pity that the mime type of typescript has not been determined yet. Maybe we can do special processing for .ts files.

File: std/http/file_server.ts [hunk 2]
```diff
@@ -99,25 +72,7 @@ function fileLenToString(len: number): string {
...
-    isDir,
-    mode
-  )}</td><td>${sizeStr}</td><td><a href="${url}">${name}${
-    isDir ? "/" : ""
-  }</a></td>
-  </tr>
-  `;
-}
-
+// TODO: load files using template
```
Comment (ry) [inline std/http/file_server.ts]: What does this mean? I would opt that you remove this TODO.

File: std/http/file_server.ts (hunk 1 above)
Comment (ry) [reply std/http/file_server.ts]: Let's go with the behavior @zhmushan has changed for now - it's quite annoying when browsing files that they get downloaded...

File: std/http/template.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
+export interface EntryInfo {
```
Comment (ry) [inline std/http/template.ts]: Could you merge into file_server.ts ?

**Comment by axetroy** [comment]
If you want to support custom templates

Consider defining a template interface

As long as the template that meets this interface can be used

eg.

```ts
interface Template {
  // defined render
  renderFile(): string
  renderFolder(): string
}

class CustomerTemplate implements Template {
  renderFile() {
    //
  }
  renderFolder() {
    //
  }
}
```

The interface need to be discussed

May be a bit complicated, but it can be highly abstract

File: std/http/file_server.ts (hunk 2 above)
Comment (zhmushan) [reply std/http/file_server.ts]: if we don’t need to render markdown/code pages, I will remove this todo

File: std/http/template.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
+import { EntryInfo } from "./file_server.ts";
```
Comment (ry) [inline std/http/template.ts]: @zhmushan Please merge this file into file_server.ts - I don't think the extra module is needed.

Comment (zhmushan) [reply std/http/template.ts]: @ry done

**Comment by ry** [comment]
This change reduced the size of the file_server bundle by half! I suppose that's because it no longer needs the content-type database?

Screen Shot 2019-12-07 at 4 27 22 PM

**Pull request merged**